zerver/actions/message_send.py

```diff
@@ -1570,53 +1575,56 @@ def check_can_send_direct_message(
     if all(user_profile.is_bot or user_profile.id == sender.id for user_profile in recipient_users):
         return
 
-    direct_message_permission_group = realm.direct_message_permission_group
-
-    if (
-        not hasattr(direct_message_permission_group, "named_user_group")
-        or direct_message_permission_group.named_user_group.name != SystemGroups.EVERYONE
-    ):
-        user_ids = [recipient_user.id for recipient_user in recipient_users] + [sender.id]
-        if not is_any_user_in_group(direct_message_permission_group.id, user_ids):
+    system_groups_name_dict = get_realm_system_groups_name_dict(realm.id)
+    if realm.direct_message_permission_group_id in system_groups_name_dict:
+        users = [*recipient_users, sender]
+        if not check_any_user_has_permission_by_role(
+            users, realm.direct_message_permission_group_id, system_groups_name_dict
+        ):
             is_nobody_group = (
-                hasattr(direct_message_permission_group, "named_user_group")
-                and direct_message_permission_group.named_user_group.name == SystemGroups.NOBODY
+                system_groups_name_dict[realm.direct_message_permission_group_id]
+                == SystemGroups.NOBODY
             )
             raise DirectMessagePermissionError(is_nobody_group)
+    else:
+        user_ids = [recipient_user.id for recipient_user in recipient_users] + [sender.id]
```

Comment:
It would be better to use sets here, since then one doesn't have to think about whether the sender could be in the recipient list, such as in self DMs?